File: 014.py
#!/usr/bin/python3

import unittest
from functools import reduce

def longest_common_prefix(words):
    def longest_prefix_2(left, right):
        pre = ""
        for (e, f) in zip(left, right):
            if e == f:
                pre += e
            else:
                break
        return pre

    if not words:
        return ""
    elif len(words) == 1:
        return words[0]

    return reduce(longest_prefix_2, words)


class test(unittest.TestCase):

    def test1(self):
        self.assertEqual(longest_common_prefix(["dog", "cat"]), "")

    def test2(self):
        self.assertEqual(longest_common_prefix(["dog", "dog"]), "dog")

    def test3(self):
        self.assertEqual(longest_common_prefix(["gus", "gussy", "gud"]), "gu")

if __name__ == '__main__':
    unittest.main()

File: 014.scala
import scala.annotation.tailrec

def longestCommonPrefix(words: List[String]): String = {
  def compareWords(left: String, right: String): String = {
    @tailrec
    def acc(pre: String, left: String, right: String): String = {
      (left.length, right.length) match {
        case (x, y) if x > 0 && y > 0 && left.charAt(0) == right.charAt(0) => acc(pre + left.charAt(0), left.substring(1), right.substring(1))
        case (_, _) => pre
      }
    }
    acc("", left, right)
  }
  words match {
    case Nil => ""
    case _ => words.tail.foldLeft(words.head)(compareWords)
  }
}
